main/src/main/scala/sbt/Keys.scala

```diff
@@ -162,8 +162,8 @@ object Keys {
 
   // Output paths
   val classDirectory = SettingKey[File]("class-directory", "Directory for compiled classes and copied resources.", AMinusSetting)
-  val cleanFiles = SettingKey[Seq[File]]("clean-files", "The files to recursively delete during a clean.", BSetting)
-  val cleanKeepFiles = SettingKey[Seq[File]]("clean-keep-files", "Files to keep during a clean.", CSetting)
+  val cleanFiles = TaskKey[Seq[File]]("clean-files", "The files to recursively delete during a clean.", BSetting)
```

Comment:
In general setting should be mostly simple values, and behaviors and things that's based on disk and network should be tasks.
The list of exact thing that we are going to clean I think should be a task for both reasons:
1. Thing that needs deleting depends on the state of the disk.
2. If we try to let the `"streams"` directory survive, or do something like https://github.com/sbt/sbt/issues/106 (let the directory for classpath survive), it's really complicated to describe it in terms of a setting.

It's true that benefit of the setting is that becomes queryable, but I don't think there's a String encoding for it, and effectively it's a Function1 wrapper that's not useful from the IDE side. Task that can tell you exactly what needs to be deleted I think provides us opportunity to fine tune the semantics of `clean` down the line, and makes the implementation of `clean` really simple.